kpatch-build/create-diff-object.c

```diff
@@ -1086,193 +1086,206 @@ static struct symbol *kpatch_find_static_twin(struct section *sec,
+ *   same between the base and patched objects, as there's no
+ *   other way to distinguish them.)
+ *
+ * - Static locals are usually referenced by functions, but
+ *   they can occasionally be referenced by data sections as
+ *   well.
  */
 void kpatch_correlate_static_local_variables(struct kpatch_elf *base,
 					     struct kpatch_elf *patched)
 {
-	struct symbol *sym, *patched_sym, *tmpsym;
+	struct symbol *sym, *patched_sym;
 	struct section *sec;
-	struct rela *rela;
-	int bundled, patched_bundled;
+	struct rela *rela, *rela2;
+	int bundled, patched_bundled, found;
 
+	/*
+	 * First undo the correlations for all static locals.  Two static
+	 * locals can have the same numbered suffix in the base and patched
+	 * objects by coincidence.
+	 */
 	list_for_each_entry(sym, &base->symbols, list) {
 
-		if (sym->type != STT_OBJECT || sym->bind != STB_LOCAL)
+		if (!kpatch_is_normal_static_local(sym))
 			continue;
 
-		if (!strchr(sym->name, '.'))
-			continue;
-
-		if (is_special_static(sym))
-			continue;
-
-		/*
-		 * Undo any previous correlation.  Two static locals can have
-		 * the same numbered suffix in the base and patched objects by
-		 * coincidence.
-		 */
 		if (sym->twin) {
 			sym->twin->twin = NULL;
 			sym->twin = NULL;
 		}
+
 		bundled = sym == sym->sec->sym;
 		if (bundled && sym->sec->twin) {
 			sym->sec->twin->twin = NULL;
 			sym->sec->twin = NULL;
 		}
+	}
 
-		/*
-		 *
-		 * Some surprising facts about static local variable symbols:
-		 *
-		 * - It's possible for multiple functions to use the same
-		 *   static local variable if the variable is defined in an
-		 *   inlined function.
-		 *
-		 * - It's also possible for multiple static local variables
-		 *   with the same name to be used in the same function if they
-		 *   have different scopes.  (We have to assume that in such
-		 *   cases, the order in which they're referenced remains the
-		 *   same between the base and patched objects, as there's no
-		 *   other way to distinguish them.)
-		 *
-		 * - Static locals are usually referenced by functions, but
-		 *   they can occasionally be referenced by data sections as
-		 *   well.
-		 *
-		 * For each section which references the variable in the base
-		 * object, look for a corresponding reference in the section's
-		 * twin in the patched object.
-		 */
-		patched_sym = NULL;
-		list_for_each_entry(sec, &base->sections, list) {
+	/*
+	 * Do the correlations: for each section reference to a static local,
+	 * look for a corresponding reference in the section's twin.
+	 */
+	list_for_each_entry(sec, &base->sections, list) {
+
+		if (!is_rela_section(sec) ||
+		    is_debug_section(sec))
+			continue;
 
-			if (!is_rela_section(sec) ||
-			    is_debug_section(sec))
+		list_for_each_entry(rela, &sec->relas, list) {
+
+			sym = rela->sym;
+			if (!kpatch_is_normal_static_local(sym))
 				continue;
 
-			list_for_each_entry(rela, &sec->relas, list) {
+			if (sym->twin)
+				continue;
```

Comment:
i'm not seeing how this case can be hit.  we set all the normal_static_local sym->twin = NULL in the previous code block.  there doesn't seem do be a way for sym->twin to get set out of order.  am i missing something?